pkg/kstatus/observe/reader/caching_reader.go

```diff
@@ -0,0 +1,225 @@
+	// of GVK and namespace. Before each polling cycle, the framework will call the
+	// Sync function, which is responsible for repopulating the cache.
+	cache map[gvkNamespace]unstructured.UnstructuredList
+}
+
+// gvkNamespace contains information about a GroupVersionKind and a namespace.
+type gvkNamespace struct {
+	GVK       schema.GroupVersionKind
+	Namespace string
+}
+
+// Get looks up the resource identified by the key and the object GVK in the cache. If the needed combination
+// of GVK and namespace is not part of the cache, that is considered an error.
+func (c *CachingClusterReader) Get(_ context.Context, key client.ObjectKey, obj *unstructured.Unstructured) error {
+	c.RLock()
+	defer c.RUnlock()
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	mapping, err := c.mapper.RESTMapping(gvk.GroupKind())
+	if err != nil {
+		return err
+	}
+	gn := gvkNamespace{
+		GVK:       gvk,
+		Namespace: key.Namespace,
+	}
+	cacheList, found := c.cache[gn]
+	if !found {
+		return fmt.Errorf("GVK %s and Namespace %s not found in cache", gvk.String(), gn.Namespace)
+	}
+	for _, u := range cacheList.Items {
+		if u.GetName() == key.Name {
+			obj.Object = u.Object
+			return nil
+		}
+	}
+	return errors.NewNotFound(mapping.Resource.GroupResource(), key.Name)
+}
+
+// ListNamespaceScoped lists all resource identifier by the GVK of the list, the namespace and the selector
+// from the cache. If the needed combination of GVK and namespace is not part of the cache, that is considered an error.
+func (c *CachingClusterReader) ListNamespaceScoped(_ context.Context, list *unstructured.UnstructuredList, namespace string, selector labels.Selector) error {
+	c.RLock()
+	defer c.RUnlock()
+	gvk := list.GroupVersionKind()
+	gn := gvkNamespace{
+		GVK:       gvk,
+		Namespace: namespace,
+	}
+
+	cacheList, found := c.cache[gn]
+	if !found {
+		return fmt.Errorf("GVK %s and Namespace %s not found in cache", gvk.String(), gn.Namespace)
+	}
+
+	var items []unstructured.Unstructured
+	for _, u := range cacheList.Items {
+		if selector.Matches(labels.Set(u.GetLabels())) {
+			items = append(items, u)
+		}
+	}
+	list.Items = items
+	return nil
+}
+
+// ListClusterScoped lists all resource identifier by the GVK of the list and selector
+// from the cache. If the needed combination of GVK and namespace (which for clusterscoped resources
+// will always be the empty string) is not part of the cache, that is considered an error.
+func (c *CachingClusterReader) ListClusterScoped(ctx context.Context, list *unstructured.UnstructuredList, selector labels.Selector) error {
+	return c.ListNamespaceScoped(ctx, list, "", selector)
+}
+
+// Sync loops over the list of gvkNamespace we know of, and uses list calls to fetch the resources.
+// This information populates the cache.
+func (c *CachingClusterReader) Sync(ctx context.Context) error {
+	c.Lock()
+	defer c.Unlock()
+	cache := make(map[gvkNamespace]unstructured.UnstructuredList)
+	for _, gn := range c.gns {
```

Comment:
FWIW, our implementation parallelizes these to speed up the sync (and locks the cache at the ns-gvk level instead of globally). We also retry server errors and client timeouts a few times (with backoff) on the list requests, since empty caches are problematic for the polling cycle when using this strategy (though perhaps less critically so in your implementation, I'm not sure). In an older implementation, we let the equivalent of the cluster reader fall back on making individual requests when the required cache was empty, but I think that made less sense than just retrying the request we actually wanted to make.

Reply:
I have considered running this in multiple goroutines, but I consider that as a more advanced feature. I'm not sure if a more fine-grained locking scheme here would make a significant difference, in particular since it is already using Read-Write locks, so it should already allow multiple readers.
Retry is a good point. I think the client already does some selective retries based on the error code returned from the server. I'll have to check with the kubebuilder team on this. I don't want multiple layers of retry.

Reply:
>  I consider that as a more advanced feature

Do you mean this is something you'd add later in the development of the library if practical feedback warrants it, or would this be something library users would have to implement their own reader for?

Reply:
Certainly something we could implement as part of the library. Just doesn't seem like the highest priority right now. Shouldn't be too hard to implement a workqueue to do this based on a go channel and goroutines.

This is something that would actually be hard for users to change since it will require changes in the core library loop.